src/hasSubObject.ts

```diff
@@ -1,7 +1,12 @@
-import type { Simplify } from "type-fest";
+import type { Merge } from "type-fest";
+import type { Brand } from "./internal/types";
 import { isDeepEqual } from "./isDeepEqual";
 import { purry } from "./purry";
 
+declare const hasSubObjectBrandId: unique symbol;
+
+export type HasSubObjectBrand = Brand<typeof hasSubObjectBrandId>;
```

Comment:
1. The name feels off to me. Why is it called "has"? The "has" prefix should be used for boolean values.
2. Don't export the type here, we don't include types in our exports as we don't consider our types part of the offering. See my comment in the test file, you don't need this exported.
3. Let's pull more of the common type up so it's easier to follow the logic. Taking my comment from "types.ts" about calling the utility type Branded.
```suggestion
declare const BRAND_SUB_OBJECT: unique symbol;

type SubObject<T, S> = Branded<Merge<T, S>, typeof BRAND_SUB_OBJECT>;
```

Now functions need to return `is SubObject<T, S>` which fully covers the intent of the type without having to understand the branding concept.

Reply:
@eranhirsch 

1. The naming has this format `<functionName><name>` and since function is called `hasSubObject` it has the name `HasSubObjectBrand`.
2. 👌
3. 👌

Reply:
The naming confused me but you are completely right that the Has part comes from the function name 🤦🏼 , my bad.

Reply:
Done.

Decided to keep the convention for naming `<functionName><name>` to make it easier to identify where this type comes from for the users. Using `SubObject` name might be confusing because the type guard doesn't resolve to sub-object but to a merged super-object and sub-object. Used a generic name `HasSubObjectGuard` instead to save connection of the type to that function and show that it's a type from a type guard function. Other options:

- `HasSubObjectData`
- `HasSubObjectMergedData`
- `HasSubObjectNarrowedData`
- `MergedData` or `MergedObject`
- `NarrowedData` or `NarrowedObject`
- `DataWithSubObject`
- `DataHasSubObject`

Let me know if I should rename it and I will do it.

By the way, wrapping it in a separate type makes it less convenient to see the type information from TS for the users:

| After | Before |
| ---- | ---- |
| ![image](https://github.com/remeda/remeda/assets/10157660/f79898de-3ebf-4945-837e-81c5c303148a) | ![image](https://github.com/remeda/remeda/assets/10157660/bf246535-5328-45fa-9457-a5eaacda0410)

It's not a big deal, just letting you know.